### PR TITLE
Make sure to not restart application-muted microphone tracks in case the audio unit gets unsuspended

### PR DESCRIPTION
**Pull request by youennf**
#### 439677029dc8350cb7a128fa96cc51f8b0c59496
<pre>
Make sure to not restart application-muted microphone tracks in case the audio unit gets unsuspended
<a href="https://bugs.webkit.org/show_bug.cgi?id=243248">https://bugs.webkit.org/show_bug.cgi?id=243248</a>
rdar://97612254

BaseAudioSharedUnit unmuted all sources currently.
This overrides sources that were muted by the application.

To prevent this, at the time the unit gets interrupted, we mute the sources but store whether these sources were producing data.
Then, when unsuspending, we only unmute sources that were producing data.

To be able to write a test, we move the interruption code from iOS to all platforms.
We add IPC to allow Internals to trigger interruptions in GPUProcess through IPC.

Covered by newly added test LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption.html.

Reviewed by Eric Carlson.

* LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption-expected.txt: Added.
* LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption.html: Added.
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::beginInterruption):
(WebCore::AudioSession::endInterruption):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebCore/platform/mediastream/ios/CoreAudioCaptureSourceIOS.h:
* Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp:
(WebCore::BaseAudioSharedUnit::addClient):
(WebCore::BaseAudioSharedUnit::removeClient):
(WebCore::BaseAudioSharedUnit::clearClients):
(WebCore::BaseAudioSharedUnit::resume):
(WebCore::BaseAudioSharedUnit::suspend):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
(WebCore::CoreAudioCaptureSource::create):
(WebCore::CoreAudioCaptureSource::createForTesting):
(WebCore::CoreAudioCaptureSourceFactory::CoreAudioCaptureSourceFactory):
(WebCore::CoreAudioCaptureSourceFactory::~CoreAudioCaptureSourceFactory):
(WebCore::CoreAudioCaptureSourceFactory::unit):
(WebCore::CoreAudioCaptureSourceFactory::beginInterruption):
(WebCore::CoreAudioCaptureSourceFactory::endInterruption):
(WebCore::CoreAudioCaptureSource::startProducingData):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
(WebCore::CoreAudioCaptureSource::canResumeAfterInterruption const):
(WebCore::CoreAudioCaptureSource::setCanResumeAfterInterruption):
(WebCore::CoreAudioCaptureSourceFactory::setOverrideUnit):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::beginAudioSessionInterruption):
(WebCore::Internals::endAudioSessionInterruption):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::triggerBeginInterruptionForTesting):
(WebKit::RemoteAudioSessionProxy::triggerEndInterruptionForTesting):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.h:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::beginInterruptionForTesting):
(WebKit::RemoteAudioSession::endInterruptionForTesting):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:

Canonical link: <a href="https://commits.webkit.org/252905@main">https://commits.webkit.org/252905@main</a>
</pre>

### CODE REVIEW

File: Source/WebCore/platform/audio/AudioSession.cpp [hunk 1]
```diff
@@ -124,6 +124,8 @@ void AudioSession::removeInterruptionObserver(InterruptionObserver& observer)
 
 void AudioSession::beginInterruption()
 {
+    RELEASE_LOG_ERROR(WebRTC, "AudioSession::beginInterruption %u", m_interruptionObservers.computeSize());
```
Comment (eric-carlson) [inline Source/WebCore/platform/audio/AudioSession.cpp]: Should this really be logged as an error?

File: Source/WebCore/platform/audio/AudioSession.cpp [hunk 2]
```diff
@@ -135,6 +137,8 @@ void AudioSession::beginInterruption()
 
 void AudioSession::endInterruption(MayResume mayResume)
 {
+    RELEASE_LOG_ERROR(WebRTC, "AudioSession::endInterruption");
```
Comment (eric-carlson) [inline Source/WebCore/platform/audio/AudioSession.cpp]: Ditto

File: Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp [hunk 3]
```diff
@@ -44,6 +44,8 @@ BaseAudioSharedUnit::BaseAudioSharedUnit()
 
 void BaseAudioSharedUnit::addClient(CoreAudioCaptureSource& client)
 {
+    RELEASE_LOG_ERROR(WebRTC, "BaseAudioSharedUnit::addClient %p", this);
```
Comment (eric-carlson) [inline Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp]: Ditto

File: Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp [hunk 4]
```diff
@@ -52,6 +54,8 @@ void BaseAudioSharedUnit::addClient(CoreAudioCaptureSource& client)
 
 void BaseAudioSharedUnit::removeClient(CoreAudioCaptureSource& client)
 {
+    RELEASE_LOG_ERROR(WebRTC, "BaseAudioSharedUnit::removeClient %p", this);
```
Comment (eric-carlson) [inline Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp]: Ditto

File: Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp [hunk 5]
```diff
@@ -60,6 +64,8 @@ void BaseAudioSharedUnit::removeClient(CoreAudioCaptureSource& client)
 
 void BaseAudioSharedUnit::clearClients()
 {
+    RELEASE_LOG_ERROR(WebRTC, "BaseAudioSharedUnit::clearClients %p", this);
```
Comment (eric-carlson) [inline Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp]: Ditto

**Review by eric-carlson — APPROVED** [review]

File: Source/WebCore/platform/audio/AudioSession.cpp (hunk 1 above)
Comment (youennf) [reply Source/WebCore/platform/audio/AudioSession.cpp]: This should be removed, I did it in the new patch

File: Source/WebCore/platform/audio/AudioSession.cpp (hunk 2 above)
Comment (youennf) [reply Source/WebCore/platform/audio/AudioSession.cpp]: Ditto

Comment (youennf) [reply Source/WebCore/platform/audio/AudioSession.cpp]: This should be removed, I did it in the new patch

File: Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp (hunk 3 above)
Comment (youennf) [reply Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp]: This should be removed, I did it in the new patch

File: Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp (hunk 5 above)
Comment (youennf) [reply Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp]: This should be removed, I did it in the new patch

File: Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp (hunk 4 above)
Comment (youennf) [reply Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp]: This should be removed, I did it in the new patch

**Pull request merged**